### PR TITLE
Add function to find base stops.

### PR DESCRIPTION
**Pull request by siarheishut**
Add function to find base stops and unit-tests for it.

### CODE REVIEW

File: src/coords_converter.cpp [hunk 1]
```diff
@@ -29,6 +30,49 @@ SortStops(const renderer_utils::Stops &stops, SortMode mode) {
   return layers;
 }
 
+unordered_set <string_view>
+BaseRepeatedWithinRoute(const renderer_utils::Buses &buses) {
```
Comment (foxfrikses) [inline src/coords_converter.cpp:34]: I don't think that you need to have `Base` prefix in any of those functions. Because What `BaseBorders` means? — it's just `Borders`

File: src/coords_converter.cpp [hunk 2]
```diff
@@ -29,6 +30,49 @@ SortStops(const renderer_utils::Stops &stops, SortMode mode) {
...
+  for (auto &[_, route] : buses) {
+    base_stops.insert(route.route.front());
+    if (!route.is_roundtrip)
+      base_stops.insert(route.route.back());
+  }
+  return base_stops;
+}
+
+unordered_set <string_view>
+BaseIntersections(const renderer_utils::Buses &buses) {
```
Comment (foxfrikses) [inline src/coords_converter.cpp:62]: How about implementing these functions in a way that they'll be iterator-based?

```
void BaseRepeatedWithinRoute(const renderer_utils::Buses &buses, OutputIt dest);
void BaseBorders(const renderer_utils::Buses &buses, OutputIt dest);
void BaseIntersections(const renderer_utils::Buses &buses, OutputIt dest);
```

Comment (siarheishut) [reply src/coords_converter.cpp:62]: But functions return unordered_set, so we can't make them iterator-based, can we? And we can't change it to vector, because we'll need to use find a lot of times.

File: src/coords_converter.cpp (hunk 1 above)
Comment (siarheishut) [reply src/coords_converter.cpp:34]: Done.

File: src/coords_converter.cpp [hunk 3]
```diff
@@ -29,6 +30,49 @@ SortStops(const renderer_utils::Stops &stops, SortMode mode) {
...
+unordered_set <string_view>
+RepeatedWithinRoute(const renderer_utils::Buses &buses) {
+  unordered_set < string_view > base_stops;
+  for (auto &[_, route] : buses) {
+    unordered_map<string_view, int> stop_freqs;
+    for (auto &stop : route.route)
+      stop_freqs[stop] += (route.is_roundtrip ? 1 : 2);
+
+    for (auto &[stop, freq] : stop_freqs) {
+      if (freq > 2) {
```
Comment (foxfrikses) [inline src/coords_converter.cpp:42]: This 2 is a magic number. Nobody would ever guess (looking on the signature of the function) what exactly it does. I propose extracting this to a parameter of the function and writing a good comment explaining what this function does.

File: src/coords_converter.cpp [hunk 4]
```diff
@@ -29,6 +30,49 @@ SortStops(const renderer_utils::Stops &stops, SortMode mode) {
   return layers;
 }
 
+unordered_set <string_view>
+RepeatedWithinRoute(const renderer_utils::Buses &buses) {
```
Comment (foxfrikses) [inline src/coords_converter.cpp:34]: Doesn't this function's logic seem similar to the `Intersections`' logic? Maybe you would rename them to `IntersectionsWithinRoute` and `IntersectionsCrossRoute`? Or something similar?

File: src/coords_converter.cpp (hunk 2 above)
Comment (foxfrikses) [reply src/coords_converter.cpp]: Ack, okay, you can leave it as it is.

File: src/map_renderer.cpp [hunk 5]
```diff
@@ -66,7 +70,16 @@ CombineCoords(
...
+unordered_set<string_view> CombineBaseStops(
+    BaseStops borders, BaseStops repeated, BaseStops intersections) {
+  unordered_set < string_view > base_stops;
+  for (auto stop : borders) base_stops.insert(stop);
+  for (auto stop : repeated) base_stops.insert(stop);
+  for (auto stop : intersections) base_stops.insert(stop);
+  return base_stops;
+}
+
+unordered_map<string_view, svg::Point> TransformCoords(
```
Comment (foxfrikses) [inline src/map_renderer.cpp:82]: Please extract all the logic of map_renderer to a separate commit.

File: src/map_renderer.cpp [hunk 6]
```diff
@@ -66,7 +70,16 @@ CombineCoords(
   return result;
 }
 
-std::unordered_map<std::string_view, svg::Point> TransformCoords(
+unordered_set<string_view> CombineBaseStops(
```
Comment (foxfrikses) [inline src/map_renderer.cpp:73]: You can write a variadic template to do this job and simply name it `Combine`

File: src/map_renderer.cpp [hunk 7]
```diff
@@ -66,7 +70,16 @@ CombineCoords(
...
+    BaseStops borders, BaseStops repeated, BaseStops intersections) {
```
Comment (foxfrikses) [inline src/map_renderer.cpp:74]: These shouldn't be passed by value.

File: tests/coords_converter_test.cpp [hunk 8]
```diff
@@ -217,3 +217,65 @@ TEST(TestSpreadStops, TestSpreadStops) {
...
+
+  std::vector<TestCase> test_cases{
+      TestCase{
+          .name = "Roundtrip",
+          .buses = {{"bus1", {{"a", "b", "c", "a"}, true}}},
+          .want = {"a"}
+      },
+      TestCase{
+          .name = "Non-roundtrip",
+          .buses = {{"bus1", {{"a", "b", "c", "d"}, false}}},
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:236]: Can a non-roundtrip end on the same stop as the one it starter from?

File: tests/coords_converter_test.cpp [hunk 9]
```diff
@@ -217,3 +217,65 @@ TEST(TestSpreadStops, TestSpreadStops) {
     EXPECT_EQ(want, got) << name;
   }
 }
+
+TEST(TestBaseBorders, TestBaseBorders) {
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:221]: `Base` shouldn't be present in the names here

File: tests/coords_converter_test.cpp [hunk 10]
```diff
@@ -217,3 +217,65 @@ TEST(TestSpreadStops, TestSpreadStops) {
...
+  };
+
+  std::vector<TestCase> test_cases{
+      TestCase{
+          .name = "Roundtrip",
+          .buses = {{"bus1", {{"a", "b", "c", "b", "a", "b", "a"}, true}}},
+          .want = {"a", "b"},
+      },
+      TestCase{
+          .name = "Roundtrip",
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:261]: Non

File: tests/coords_converter_test.cpp [hunk 11]
```diff
@@ -217,3 +217,65 @@ TEST(TestSpreadStops, TestSpreadStops) {
...
+    EXPECT_EQ(want, got) << name;
+  }
+}
+
+TEST(TestBaseIntersections, TestBaseIntersections) {
+  std::unordered_set < std::string_view > want = {"c"};
+  std::unordered_set < std::string_view > got =
+      rm::coords_converter::Intersections(
+          {
+              {"bus1", {{"a", "b", "c", "a"}, true}},
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:278]: Add a few more routes to this test. One route that doesn't intersect anything. Plus, additionally to a stop being a crossroad of two routes, make some stops to be crossroads of 3+ routes (a potential edge case).

Plus please add a test for empty routes and for routes with no intersections at all.

File: src/coords_converter.cpp (hunk 4 above)
Comment (siarheishut) [reply src/coords_converter.cpp:34]: What do you think about `IntersectionsWithinRoute` and `RoutesIntersections`?

File: src/coords_converter.cpp (hunk 3 above)
Comment (siarheishut) [reply src/coords_converter.cpp:42]: Done. Is my description enough?

File: tests/coords_converter_test.cpp (hunk 9 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp:221]: Done.

File: tests/coords_converter_test.cpp (hunk 8 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp:236]: Yes, but should I check this case in this function? I don't think it needs to be tested here.

File: tests/coords_converter_test.cpp (hunk 10 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp:261]: Done.

File: src/map_renderer.cpp (hunk 5 above)
Comment (siarheishut) [reply src/map_renderer.cpp:82]: I'll make a separate commit with map_renderer logic after merging this branch and adding `Interpolate` function as well.

File: tests/coords_converter_test.cpp (hunk 11 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp:278]: Done.

File: tests/coords_converter_test.cpp (hunk 8 above)
Comment (foxfrikses) [reply tests/coords_converter_test.cpp:236]: I would?

File: src/coords_converter.cpp (hunk 4 above)
Comment (foxfrikses) [reply src/coords_converter.cpp]: The reason I proposed `IntersectionsCrossRoute` is because it gives clarity about what the function does and has the same name structure `IntersectionsWithinRoute` of [Intersections]+[of what type], which leaves no questions on their differences. Whereas RouteIntersections is a completely different name structure, which immediately raises a question "what is the difference between `RouteIntersections` and `IntersectionsWithinRoute`"

File: src/map_renderer.cpp (hunk 5 above)
Comment (foxfrikses) [reply src/map_renderer.cpp]: Cool

File: src/coords_converter.cpp [hunk 12]
```diff
@@ -23,15 +24,59 @@ SortStops(const renderer_utils::Stops &stops, SortMode mode) {
             });
   sort(begin(sorted_stops), end(sorted_stops));
 
-  vector<string_view> layers;
+  vector <string_view> layers;
```
Comment (foxfrikses) [inline src/coords_converter.cpp:27]: Are you sure that you haven't changed your style? Just check the diff of this file: it didn't put this space before `<>` ever before.

File: src/coords_converter.cpp [hunk 13]
```diff
@@ -23,15 +24,59 @@ SortStops(const renderer_utils::Stops &stops, SortMode mode) {
...
   transform(begin(sorted_stops), end(sorted_stops), back_inserter(layers),
             [](auto &item) { return item.second; });
   return layers;
 }
 
+unordered_set <string_view>
+IntersectionsWithinRoute(const renderer_utils::Buses &buses,
+                         int encounter_count) {
```
Comment (foxfrikses) [inline src/coords_converter.cpp:35]: `encounters`

File: src/coords_converter.h [hunk 14]
```diff
@@ -19,6 +20,18 @@ struct Frame { double width, height, padding; };
...
 
+std::unordered_set<std::string_view>
+IntersectionsWithinRoute(const renderer_utils::Buses &buses,
+                         int encounter_count);
+
+std::unordered_set<std::string_view>
+EndPoints(const renderer_utils::Buses &buses);
+
+// Finds stops that were encountered more than `encounter_count` times
+// within the same route.
```
Comment (foxfrikses) [inline src/coords_converter.h:31]: Btw that's what I'm talking about. You wrote a comment to a wrong function as their names are misleading.

Comment (foxfrikses) [inline src/coords_converter.h:31]: I would do this not `more than` but `at least`, as it's easier to not mess up.

`IntersectionsWithinRoute returns stops that are present in a route at least `count` times`

How about that?

Comment (siarheishut) [reply src/coords_converter.h:31]: I'd rather use:
IntersectionsWithinRoute returns stops that are present in any route at least `count` times.
Because it seems to me, that `a route` is confusing a little, am I wrong?

File: tests/coords_converter_test.cpp (hunk 8 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp]: Added.

File: src/coords_converter.cpp (hunk 13 above)
Comment (siarheishut) [reply src/coords_converter.cpp]: Done.

File: src/coords_converter.h (hunk 14 above)
Comment (siarheishut) [reply src/coords_converter.h]: Done.

File: src/coords_converter.cpp (hunk 4 above)
Comment (siarheishut) [reply src/coords_converter.cpp]: Ok. Changed to `IntersectionsCrossRoute`.

File: src/coords_converter.cpp (hunk 12 above)
Comment (siarheishut) [reply src/coords_converter.cpp]: I reset style but there are anyway some extra spaces. I don't understand how to fix it yet.

File: src/coords_converter.cpp [hunk 15]
```diff
@@ -45,8 +87,7 @@ AdjacentStops(const renderer_utils::Buses &buses) {
 
 vector <vector<string_view>> CompressNonadjacent(
     const vector <string_view> &stops,
-    const vector <pair<string_view,
-                       string_view>> &neighbors) {
+    const vector <pair<string_view, string_view>> &neighbors) {
```
Comment (foxfrikses) [inline src/coords_converter.cpp:90]: Not for this commit

File: tests/coords_converter_test.cpp [hunk 16]
```diff
@@ -217,3 +217,93 @@ TEST(TestSpreadStops, TestSpreadStops) {
...
+    auto got = rm::coords_converter::EndPoints(buses);
+    EXPECT_EQ(want, got) << name;
+  }
+}
+
+TEST(TestIntersectionsWithinRoute, TestIntersectionsWithinRoute) {
+  struct TestCase {
+    std::string name;
+    rm::renderer_utils::Buses buses;
+    std::unordered_set<std::string_view> want;
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:256]: I would add more tests: for the same buses show how the result changes with `count` progressing from 1 to 4, say, where at 1 we have all the buses, at 2 only a few, at 3 — even less, at 4 – none.

You can replace `want` with:
```
struct Subcase {
  int count;
  std::unordered_set<std::string_view> want;
};
std::vector<Subcase> subcases;
```

I would also test with multiple buses, not with one.

File: tests/coords_converter_test.cpp [hunk 17]
```diff
@@ -217,3 +217,93 @@ TEST(TestSpreadStops, TestSpreadStops) {
...
+
+  vector<TestCase> test_cases = {
+      TestCase{
+          .name = "Base request",
+          .buses = {
+              {"bus1", {{"a", "b", "c", "a"}, true}},
+              {"bus2", {{"d", "f", "c", "g"}, false}},
+              {"bus3", {{"k", "b", "c", "f"}, false}},
+              {"bus4", {{"x", "y", "z"}, false}}},
+          .want = {"c", "b", "c", "f"}
```
Comment (foxfrikses) [inline tests/coords_converter_test.cpp:293]: Why "c" twice?

File: src/coords_converter.cpp (hunk 15 above)
Comment (siarheishut) [reply src/coords_converter.cpp:90]: Fixed.

File: tests/coords_converter_test.cpp (hunk 17 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp:293]: Fixed.

File: tests/coords_converter_test.cpp (hunk 16 above)
Comment (siarheishut) [reply tests/coords_converter_test.cpp:256]: Done.

File: src/coords_converter.cpp [hunk 18]
```diff
@@ -29,6 +30,48 @@ SortStops(const renderer_utils::Stops &stops, SortMode mode) {
   return layers;
 }
 
+unordered_set <string_view>
+IntersectionsWithinRoute(const renderer_utils::Buses &buses, int count) {
+  unordered_set < string_view > base_stops;
```
[the remark this answers is not in the document]
Comment (foxfrikses) [reply src/coords_converter.cpp:35]: If it's some bug that you can't find a fix of than I think you should just specify `std::` before the containers affected. I can see that it's never a problem when you specify `std::`

**Review by foxfrikses — APPROVED** [review]
LGTM

Though I would like to see the problem with the spaces in templaces fixed: `set<string>` or `std::set<std::string>` instead of `set < string >`,  or `set< string >`, or `set <string>`.

You can submit this PR now.
If you fix the problem above, you can submit the PR as well without additional approval.

File: src/coords_converter.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/coords_converter.cpp:35]: Done.
